File: Backend/BY.Infrastructure/Services/GoalService.cs
using BY.Core.DTOs;
using BY.Core.Entities;
using BY.Core.Enums;
using BY.Core.Interfaces;
using Microsoft.Extensions.Logging;

namespace BY.Infrastructure.Services;

public class GoalService : IGoalService
{
    private readonly IUnitOfWork _unitOfWork;
    private readonly ILogger<GoalService> _logger;

    public GoalService(IUnitOfWork unitOfWork, ILogger<GoalService> logger)
    {
        _unitOfWork = unitOfWork;
        _logger = logger;
    }

    public async Task<ApiResponse<GoalResponse>> CreateGoalAsync(Guid userId, CreateGoalRequest request)
    {
        try
        {
            // Validate input
            var validationErrors = ValidateCreateGoalRequest(request);
            if (validationErrors.Any())
            {
                return ApiResponse<GoalResponse>.FailureResult("Validation failed", validationErrors);
            }

            // Check if user exists
            var user = await _unitOfWork.Users.GetByIdAsync(userId);
            if (user == null)
            {
                return ApiResponse<GoalResponse>.FailureResult("User not found");
            }

            // Set start date to today if not provided
            var startDate = request.StartDate ?? DateTime.UtcNow.Date;
            var endDate = startDate.AddDays(request.DurationDays - 1);

            // Calculate daily stake amount
            var dailyStakeAmount = Math.Round(request.TotalStakeAmount / request.DurationDays, 2);

            // Create new goal
            var goal = new Goal
            {
                UserId = userId,
                Title = request.Title.Trim(),
                Description = request.Description?.Trim() ?? string.Empty,
                Category = request.Category,
                Status = GoalStatus.Active,
                StartDate = startDate,
                EndDate = endDate,
                DurationDays = request.DurationDays,
                TotalStakeAmount = request.TotalStakeAmount,
                DailyStakeAmount = dailyStakeAmount,
                ReminderTime = request.ReminderTime,
                ReminderMessage = request.ReminderMessage?.Trim() ?? $"Time to work on: {request.Title}",
                CurrentStreak = 0,
                LongestStreak = 0,
                SuccessfulDays = 0,
                FailedDays = 0,
                MissedDays = 0,
                TotalPaid = 0
            };

            await _unitOfWork.Goals.AddAsync(goal);
            await _unitOfWork.SaveChangesAsync();

            var goalResponse = MapToGoalResponse(goal);

            _logger.LogInformation("Goal created successfully: {GoalId} for user {UserId}", goal.Id, userId);
            return ApiResponse<GoalResponse>.SuccessResult(goalResponse, "Goal created successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating goal for user {UserId}", userId);
            return ApiResponse<GoalResponse>.FailureResult("An error occurred while creating the goal");
        }
    }

    public async Task<ApiResponse<GoalResponse>> GetGoalByIdAsync(Guid goalId, Guid userId)
    {
        try
        {
            var goal = await _unitOfWork.Goals.GetByIdAsync(goalId);
            if (goal == null)
            {
                return ApiResponse<GoalResponse>.FailureResult("Goal not found");
            }

            // Check if user owns this goal
            if (goal.UserId != userId)
            {
                return ApiResponse<GoalResponse>.FailureResult("Access denied");
            }

            var goalResponse = MapToGoalResponse(goal);
            return ApiResponse<GoalResponse>.SuccessResult(goalResponse);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving goal {GoalId} for user {UserId}", goalId, userId);
            return ApiResponse<GoalResponse>.FailureResult("An error occurred while retrieving the goal");
        }
    }

    public async Task<ApiResponse<IEnumerable<GoalResponse>>> GetUserGoalsAsync(Guid userId, GoalStatus? status = null, GoalCategory? category = null)
    {
        try
        {
            var goals = await _unitOfWork.Goals.GetUserGoalsAsync(userId);

            // Apply filters
            if (status.HasValue)
            {
                goals = goals.Where(g => g.Status == status.Value);
            }

            if (category.HasValue)
            {
                goals = goals.Where(g => g.Category == category.Value);
            }

            var goalResponses = goals.Select(MapToGoalResponse).ToList();
            return ApiResponse<IEnumerable<GoalResponse>>.SuccessResult(goalResponses);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving goals for user {UserId}", userId);
            return ApiResponse<IEnumerable<GoalResponse>>.FailureResult("An error occurred while retrieving goals");
        }
    }

    public async Task<ApiResponse<IEnumerable<GoalResponse>>> GetActiveGoalsAsync(Guid userId)
    {
        try
        {
            var activeGoals = await _unitOfWork.Goals.GetActiveGoalsAsync(userId);
            var goalResponses = activeGoals.Select(MapToGoalResponse).ToList();
            return ApiResponse<IEnumerable<GoalResponse>>.SuccessResult(goalResponses);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving active goals for user {UserId}", userId);
            return ApiResponse<IEnumerable<GoalResponse>>.FailureResult("An error occurred while retrieving active goals");
        }
    }

    public async Task<ApiResponse<GoalResponse>> UpdateGoalAsync(Guid goalId, Guid userId, UpdateGoalRequest request)
    {
        try
        {
            var goal = await _unitOfWork.Goals.GetByIdAsync(goalId);
            if (goal == null)
            {
                return ApiResponse<GoalResponse>.FailureResult("Goal not found");
            }

            // Check if user owns this goal
            if (goal.UserId != userId)
            {
                return ApiResponse<GoalResponse>.FailureResult("Access denied");
            }

            // Check if goal is active (can't update completed or failed goals)
            if (goal.Status != GoalStatus.Active && goal.Status != GoalStatus.Paused)
            {
                return ApiResponse<GoalResponse>.FailureResult("Cannot update completed or failed goals");
            }

            // Update allowed properties
            if (!string.IsNullOrWhiteSpace(request.Title))
                goal.Title = request.Title.Trim();

            if (request.Description != null)
                goal.Description = request.Description.Trim();

            if (request.ReminderTime.HasValue)
                goal.ReminderTime = request.ReminderTime.Value;

            if (!string.IsNullOrWhiteSpace(request.ReminderMessage))
                goal.ReminderMessage = request.ReminderMessage.Trim();

            await _unitOfWork.Goals.UpdateAsync(goal);
            await _unitOfWork.SaveChangesAsync();

            var goalResponse = MapToGoalResponse(goal);

            _logger.LogInformation("Goal updated successfully: {GoalId} for user {UserId}", goalId, userId);
            return ApiResponse<GoalResponse>.SuccessResult(goalResponse, "Goal updated successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating goal {GoalId} for user {UserId}", goalId, userId);
            return ApiResponse<GoalResponse>.FailureResult("An error occurred while updating the goal");
        }
    }

    public async Task<ApiResponse<bool>> DeleteGoalAsync(Guid goalId, Guid userId)
    {
        try
        {
            var goal = await _unitOfWork.Goals.GetByIdAsync(goalId);
            if (goal == null)
            {
                return ApiResponse<bool>.FailureResult("Goal not found");
            }

            // Check if user owns this goal
            if (goal.UserId != userId)
            {
                return ApiResponse<bool>.FailureResult("Access denied");
            }

            await _unitOfWork.Goals.DeleteAsync(goal);
            await _unitOfWork.SaveChangesAsync();

            _logger.LogInformation("Goal deleted successfully: {GoalId} for user {UserId}", goalId, userId);
            return ApiResponse<bool>.SuccessResult(true, "Goal deleted successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting goal {GoalId} for user {UserId}", goalId, userId);
            return ApiResponse<bool>.FailureResult("An error occurred while deleting the goal");
        }
    }

    public async Task<ApiResponse<GoalResponse>> PauseGoalAsync(Guid goalId, Guid userId)
    {
        try
        {
            var goal = await _unitOfWork.Goals.GetByIdAsync(goalId);
            if (goal == null)
            {
                return ApiResponse<GoalResponse>.FailureResult("Goal not found");
            }

            if (goal.UserId != userId)
            {
                return ApiResponse<GoalResponse>.FailureResult("Access denied");
            }

            if (goal.Status != GoalStatus.Active)
            {
                return ApiResponse<GoalResponse>.FailureResult("Only active goals can be paused");
            }

            goal.Status = GoalStatus.Paused;
            await _unitOfWork.Goals.UpdateAsync(goal);
            await _unitOfWork.SaveChangesAsync();

            var goalResponse = MapToGoalResponse(goal);
            return ApiResponse<GoalResponse>.SuccessResult(goalResponse, "Goal paused successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error pausing goal {GoalId} for user {UserId}", goalId, userId);
            return ApiResponse<GoalResponse>.FailureResult("An error occurred while pausing the goal");
        }
    }

    public async Task<ApiResponse<GoalResponse>> ResumeGoalAsync(Guid goalId, Guid userId)
    {
        try
        {
            var goal = await _unitOfWork.Goals.GetByIdAsync(goalId);
            if (goal == null)
            {
                return ApiResponse<GoalResponse>.FailureResult("Goal not found");
            }

            if (goal.UserId != userId)
            {
                return ApiResponse<GoalResponse>.FailureResult("Access denied");
            }

            if (goal.Status != GoalStatus.Paused)
            {
                return ApiResponse<GoalResponse>.FailureResult("Only paused goals can be resumed");
            }

            goal.Status = GoalStatus.Active;
            await _unitOfWork.Goals.UpdateAsync(goal);
            await _unitOfWork.SaveChangesAsync();

            var goalResponse = MapToGoalResponse(goal);
            return ApiResponse<GoalResponse>.SuccessResult(goalResponse, "Goal resumed successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error resuming goal {GoalId} for user {UserId}", goalId, userId);
            return ApiResponse<GoalResponse>.FailureResult("An error occurred while resuming the goal");
        }
    }

    public async Task<ApiResponse<CheckInResponse>> CheckInAsync(Guid goalId, Guid userId, CheckInRequest request)
    {
        try
        {
            var goal = await _unitOfWork.Goals.GetByIdAsync(goalId);
            if (goal == null)
            {
                return ApiResponse<CheckInResponse>.FailureResult("Goal not found");
            }

            if (goal.UserId != userId)
            {
                return ApiResponse<CheckInResponse>.FailureResult("Access denied");
            }

            if (goal.Status != GoalStatus.Active)
            {
                return ApiResponse<CheckInResponse>.FailureResult("Goal is not active");
            }

            var today = DateOnly.FromDateTime(DateTime.UtcNow.Date);

            // Check if already checked in today
            var existingCheckIn = await _unitOfWork.CheckIns.GetTodayCheckInAsync(goalId, today);
            if (existingCheckIn != null)
            {
                return ApiResponse<CheckInResponse>.FailureResult("Already checked in for today");
            }

            // Create check-in
            var checkIn = new CheckIn
            {
                GoalId = goalId,
                Date = today,
                Completed = request.Completed,
                Notes = request.Notes?.Trim(),
                CheckInTime = DateTime.UtcNow,
                PaymentProcessed = false,
                AmountCharged = 0,
                StreakCount = 0
            };

            // Update goal statistics
            if (request.Completed)
            {
                goal.SuccessfulDays++;
                goal.CurrentStreak++;
                if (goal.CurrentStreak > goal.LongestStreak)
                {
                    goal.LongestStreak = goal.CurrentStreak;
                }
                checkIn.StreakCount = goal.CurrentStreak;
            }
            else
            {
                goal.FailedDays++;
                goal.CurrentStreak = 0;
                goal.TotalPaid += goal.DailyStakeAmount;
                checkIn.PaymentProcessed = true;
                checkIn.AmountCharged = goal.DailyStakeAmount;
            }

            await _unitOfWork.CheckIns.AddAsync(checkIn);
            await _unitOfWork.Goals.UpdateAsync(goal);
            await _unitOfWork.SaveChangesAsync();

            var checkInResponse = MapToCheckInResponse(checkIn, request.Completed ? 
                "Great job! Keep up the momentum!" : 
                "Don't worry - tomorrow is a fresh start!");

            return ApiResponse<CheckInResponse>.SuccessResult(checkInResponse, "Check-in recorded successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error during check-in for goal {GoalId}, user {UserId}", goalId, userId);
            return ApiResponse<CheckInResponse>.FailureResult("An error occurred during check-in");
        }
    }

    public async Task<ApiResponse<IEnumerable<CheckInResponse>>> GetGoalCheckInsAsync(Guid goalId, Guid userId, int page = 1, int limit = 30)
    {
        try
        {
            var goal = await _unitOfWork.Goals.GetByIdAsync(goalId);
            if (goal == null)
            {
                return ApiResponse<IEnumerable<CheckInResponse>>.FailureResult("Goal not found");
            }

            if (goal.UserId != userId)
            {
                return ApiResponse<IEnumerable<CheckInResponse>>.FailureResult("Access denied");
            }

            var checkIns = await _unitOfWork.CheckIns.GetGoalCheckInsAsync(goalId);
            
            var paginatedCheckIns = checkIns
                .Skip((page - 1) * limit)
                .Take(limit)
                .Select(c => MapToCheckInResponse(c))
                .ToList();

            return ApiResponse<IEnumerable<CheckInResponse>>.SuccessResult(paginatedCheckIns);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving check-ins for goal {GoalId}, user {UserId}", goalId, userId);
            return ApiResponse<IEnumerable<CheckInResponse>>.FailureResult("An error occurred while retrieving check-ins");
        }
    }

    private List<string> ValidateCreateGoalRequest(CreateGoalRequest request)
    {
        var errors = new List<string>();

        if (string.IsNullOrWhiteSpace(request.Title))
            errors.Add("Title is required");
        else if (request.Title.Length > 200)
            errors.Add("Title must be less than 200 characters");

        if (request.DurationDays < 1)
            errors.Add("Duration must be at least 1 day");
        else if (request.DurationDays > 365)
            errors.Add("Duration cannot exceed 365 days");

        if (request.TotalStakeAmount < 0)
            errors.Add("Stake amount cannot be negative");

        return errors;
    }

    private static GoalResponse MapToGoalResponse(Goal goal)
    {
        var totalDays = goal.SuccessfulDays + goal.FailedDays;
        var progressPercentage = goal.DurationDays > 0 ? (decimal)totalDays / goal.DurationDays * 100 : 0;
        var successRate = totalDays > 0 ? (decimal)goal.SuccessfulDays / totalDays * 100 : 0;

        return new GoalResponse
        {
            Id = goal.Id,
            Title = goal.Title,
            Description = goal.Description,
            Category = goal.Category,
            Status = goal.Status,
            StartDate = goal.StartDate,
            EndDate = goal.EndDate,
            DurationDays = goal.DurationDays,
            TotalStakeAmount = goal.TotalStakeAmount,
            DailyStakeAmount = goal.DailyStakeAmount,
            ReminderTime = goal.ReminderTime,
            SuccessfulDays = goal.SuccessfulDays,
            FailedDays = goal.FailedDays,
            MissedDays = goal.MissedDays,
            CurrentStreak = goal.CurrentStreak,
            LongestStreak = goal.LongestStreak,
            TotalPaid = goal.TotalPaid,
            ProgressPercentage = Math.Round(progressPercentage, 1),
            SuccessRate = Math.Round(successRate, 1),
            IsActive = goal.Status == GoalStatus.Active,
            CreatedAt = goal.CreatedAt
        };
    }

    private static CheckInResponse MapToCheckInResponse(CheckIn checkIn, string? message = null)
    {
        return new CheckInResponse
        {
            Id = checkIn.Id,
            GoalId = checkIn.GoalId,
            Date = checkIn.Date,
            Completed = checkIn.Completed,
            Notes = checkIn.Notes,
            CheckInTime = checkIn.CheckInTime,
            PaymentProcessed = checkIn.PaymentProcessed,
            AmountCharged = checkIn.AmountCharged,
            StreakCount = checkIn.StreakCount,
            Message = message
        };
    }
}

File: MobileApp/src/services/goalService.ts
import { apiClient } from './apiClient';
import { 
  Goal, 
  CreateGoalRequest, 
  UpdateGoalRequest,
  CheckInRequest,
  CheckInResponse,
  GoalStatus,
  GoalCategory,
  ApiResponse 
} from '@/types';

export class GoalService {
  // Create a new goal
  async createGoal(request: CreateGoalRequest): Promise<ApiResponse<Goal>> {
    return await apiClient.post<Goal>('/goals', request);
  }

  // Get all goals for the current user
  async getUserGoals(status?: GoalStatus, category?: GoalCategory): Promise<ApiResponse<Goal[]>> {
    const params: any = {};
    if (status) params.status = status;
    if (category !== undefined) params.category = category;
    
    return await apiClient.get<Goal[]>('/goals', params);
  }

  // Get active goals for the current user
  async getActiveGoals(): Promise<ApiResponse<Goal[]>> {
    return await apiClient.get<Goal[]>('/goals/active');
  }

  // Get a specific goal by ID
  async getGoal(goalId: string): Promise<ApiResponse<Goal>> {
    return await apiClient.get<Goal>(`/goals/${goalId}`);
  }

  // Update a goal
  async updateGoal(goalId: string, request: UpdateGoalRequest): Promise<ApiResponse<Goal>> {
    return await apiClient.put<Goal>(`/goals/${goalId}`, request);
  }

  // Delete a goal
  async deleteGoal(goalId: string): Promise<ApiResponse<boolean>> {
    return await apiClient.delete<boolean>(`/goals/${goalId}`);
  }

  // Pause an active goal
  async pauseGoal(goalId: string): Promise<ApiResponse<Goal>> {
    return await apiClient.post<Goal>(`/goals/${goalId}/pause`);
  }

  // Resume a paused goal
  async resumeGoal(goalId: string): Promise<ApiResponse<Goal>> {
    return await apiClient.post<Goal>(`/goals/${goalId}/resume`);
  }

  // Check in for a goal (legacy method - prefer using CheckInService)
  async checkIn(goalId: string, request: CheckInRequest): Promise<ApiResponse<CheckInResponse>> {
    return await apiClient.post<CheckInResponse>(`/goals/${goalId}/checkin`, request);
  }

  // Get check-in history for a goal (legacy method - prefer using CheckInService)
  async getGoalCheckIns(
    goalId: string, 
    page: number = 1, 
    limit: number = 30
  ): Promise<ApiResponse<CheckInResponse[]>> {
    return await apiClient.get<CheckInResponse[]>(`/goals/${goalId}/checkins`, {
      page,
      limit,
    });
  }

  // Mobile-specific helper methods

  // Get goals by status with mobile-friendly filtering
  async getGoalsByStatus(status: GoalStatus): Promise<Goal[]> {
    try {
      const response = await this.getUserGoals(status);
      return response.success && response.data ? response.data : [];
    } catch (error) {
      console.error(`Error getting ${status} goals:`, error);
      return [];
    }
  }

  // Get goals by category
  async getGoalsByCategory(category: GoalCategory): Promise<Goal[]> {
    try {
      const response = await this.getUserGoals(undefined, category);
      return response.success && response.data ? response.data : [];
    } catch (error) {
      console.error(`Error getting goals for category ${category}:`, error);
      return [];
    }
  }

  // Get goals requiring check-in today
  async getGoalsRequiringCheckIn(): Promise<Goal[]> {
    try {
      const activeGoals = await this.getGoalsByStatus(GoalStatus.Active);
      
      // In a real app, you'd check if they've already checked in today
      // For now, return all active goals
      return activeGoals;
    } catch (error) {
      console.error('Error getting goals requiring check-in:', error);
      return [];
    }
  }

  // Get goal statistics for dashboard
  async getGoalStats(): Promise<{
    total: number;
    active: number;
    completed: number;
    failed: number;
    paused: number;
  }> {
    try {
      const response = await this.getUserGoals();
      
      if (response.success && response.data) {
        const goals = response.data;
        
        return {
          total: goals.length,
          active: goals.filter(g => g.status === GoalStatus.Active).length,
          completed: goals.filter(g => g.status === GoalStatus.Completed).length,
          failed: goals.filter(g => g.status === GoalStatus.Failed).length,
          paused: goals.filter(g => g.status === GoalStatus.Paused).length,
        };
      }
      
      return { total: 0, active: 0, completed: 0, failed: 0, paused: 0 };
    } catch (error) {
      console.error('Error getting goal stats:', error);
      return { total: 0, active: 0, completed: 0, failed: 0, paused: 0 };
    }
  }

  // Quick create goal with validation
  async quickCreateGoal(
    title: string,
    category: GoalCategory,
    durationDays: number,
    totalStakeAmount: number
  ): Promise<ApiResponse<Goal>> {
    // Basic client-side validation
    if (!title.trim()) {
      return {
        success: false,
        message: 'Title is required',
        errors: ['Title cannot be empty'],
      };
    }

    if (durationDays < 1 || durationDays > 365) {
      return {
        success: false,
        message: 'Duration must be between 1 and 365 days',
        errors: ['Invalid duration'],
      };
    }

    if (totalStakeAmount < 0.01 || totalStakeAmount > 10000) {
      return {
        success: false,
        message: 'Stake amount must be between $0.01 and $10,000',
        errors: ['Invalid stake amount'],
      };
    }

    const request: CreateGoalRequest = {
      title: title.trim(),
      description: `${title} - Created on mobile`,
      category,
      durationDays,
      totalStakeAmount,
      reminderTime: '09:00:00', // Default reminder time
      reminderMessage: `Time to work on: ${title}`,
    };

    return await this.createGoal(request);
  }

  // Get user's most successful category
  async getMostSuccessfulCategory(): Promise<GoalCategory | null> {
    try {
      const goals = await this.getGoalsByStatus(GoalStatus.Completed);
      
      if (goals.length === 0) return null;
      
      // Count goals by category
      const categoryCount = goals.reduce((acc, goal) => {
        acc[goal.category] = (acc[goal.category] || 0) + 1;
        return acc;
      }, {} as Record<GoalCategory, number>);
      
      // Find category with most completed goals
      const mostSuccessful = Object.entries(categoryCount)
        .reduce((max, [category, count]) => 
          count > max.count ? { category: Number(category) as GoalCategory, count } : max,
          { category: GoalCategory.Other, count: 0 }
        );
      
      return mostSuccessful.category;
    } catch (error) {
      console.error('Error getting most successful category:', error);
      return null;
    }
  }

  // Calculate overall success rate
  async getOverallSuccessRate(): Promise<number> {
    try {
      const response = await this.getUserGoals();
      
      if (response.success && response.data) {
        const goals = response.data;
        const completedGoals = goals.filter(g => 
          g.status === GoalStatus.Completed || g.status === GoalStatus.Failed
        );
        
        if (completedGoals.length === 0) return 0;
        
        const successfulGoals = goals.filter(g => g.status === GoalStatus.Completed).length;
        return Math.round((successfulGoals / completedGoals.length) * 100);
      }
      
      return 0;
    } catch (error) {
      console.error('Error calculating success rate:', error);
      return 0;
    }
  }

  // Get current streak across all active goals
  async getCurrentStreak(): Promise<number> {
    try {
      const activeGoals = await this.getGoalsByStatus(GoalStatus.Active);
      
      if (activeGoals.length === 0) return 0;
      
      // Return the highest current streak among active goals
      return Math.max(...activeGoals.map(g => g.currentStreak));
    } catch (error) {
      console.error('Error getting current streak:', error);
      return 0;
    }
  }

  // Simple check-in for mobile (Yes/No only)
  async simpleCheckIn(goalId: string, completed: boolean, notes?: string): Promise<ApiResponse<CheckInResponse>> {
    const request: CheckInRequest = {
      completed,
      notes,
    };
    
    return await this.checkIn(goalId, request);
  }
}

export const goalService = new GoalService();